### ISSUE TITLE
Improve error reporting when drenv completes

### ISSUE THREAD

**Issue by nirs** [RamenDR/ramen#820]
When one of the clusters fail during start, we can get confusing output:

```
$ drenv start regional-dr.yaml
logs
logs
logs
logs
Traceback about cluster failing...

logs
logs
logs
logs
logs
logs
logs
logs
logs
logs
last log line
```

The error may be hidden behind screen full of successful logs from other clusters.

I have seen 2 developers failing to detect or understand the errors:
- One thought that running again only the failed component is enough to recover
  - you need to run `drenv start` again, since when one component fails, the next components do not run
- One had running drenv without rbd-mirror - this can happen if one of the clusters failed. In this case we don't run the next part, configuring rbd-mirror.

Fix:
- [ ] Change error reporting so *all* errors are reported at the end of logs - like other errors
- [ ] Add a clear message when the entire environment was started successfully
- [ ] When an error occurred, explain how to recover from the error

I can see success message like:

    Start completed successfully!

    Status:
      hub: OK
      dr1: OK
      dr2: OK

When start failed, maybe:

    Start failed!

    Status:
      hub: OK
      dr1: ERROR
      dr2: OK

    Errors:

      Traceback from cluster dr1...
 
   To recover run `drenv start regional-dr.yaml` again.

**Comment by nirs** [comment]
Example of a random error - if you work with small terminal window, you may
miss the error since it was already scrolled out of the window.

```
$ drenv start regional-dr.yaml 
2023-03-30 19:58:55,615 INFO    [rdr] Starting environment
2023-03-30 19:58:55,663 INFO    [dr1] Starting minikube cluster
2023-03-30 19:58:56,688 INFO    [dr2] Starting minikube cluster
2023-03-30 19:58:57,684 INFO    [hub] Starting minikube cluster
2023-03-30 19:59:37,266 INFO    [dr1] Cluster started in 41.60 seconds
2023-03-30 19:59:37,267 INFO    [dr1/0] Running cert-manager/start
2023-03-30 19:59:37,267 INFO    [dr1/1] Running ocm-cluster/start
2023-03-30 19:59:37,270 INFO    [dr1/2] Running csi-addons/start
2023-03-30 19:59:54,620 INFO    [dr2] Cluster started in 57.93 seconds
2023-03-30 19:59:54,621 INFO    [dr2/0] Running cert-manager/start
2023-03-30 19:59:54,622 INFO    [dr2/1] Running ocm-cluster/start
2023-03-30 19:59:54,624 INFO    [dr2/2] Running csi-addons/start
2023-03-30 20:00:20,654 INFO    [hub] Cluster started in 82.97 seconds
2023-03-30 20:00:20,655 INFO    [hub/0] Running ocm-hub/start
2023-03-30 20:00:20,912 INFO    [dr1/0] cert-manager/start completed in 43.65 seconds
2023-03-30 20:00:20,912 INFO    [dr1/0] Running rook-operator/start
2023-03-30 20:00:30,600 INFO    [dr1/2] csi-addons/start completed in 53.33 seconds
2023-03-30 20:00:30,600 INFO    [dr1/2] Running olm/start
2023-03-30 20:00:33,743 INFO    [dr2/0] cert-manager/start completed in 39.12 seconds
2023-03-30 20:00:33,743 INFO    [dr2/0] Running rook-operator/start
2023-03-30 20:00:59,000 INFO    [dr2/2] csi-addons/start completed in 64.38 seconds
2023-03-30 20:00:59,001 INFO    [dr2/2] Running olm/start
2023-03-30 20:02:39,660 INFO    [dr1/0] rook-operator/start completed in 138.75 seconds
2023-03-30 20:02:39,660 INFO    [dr1/0] Running rook-cluster/start
2023-03-30 20:03:13,937 INFO    [dr2/0] rook-operator/start completed in 160.19 seconds
2023-03-30 20:03:13,937 INFO    [dr2/0] Running rook-cluster/start
2023-03-30 20:04:43,878 INFO    [hub/0] ocm-hub/start completed in 263.22 seconds
2023-03-30 20:04:43,878 INFO    [hub/0] Running ocm-controller/start
2023-03-30 20:05:16,959 INFO    [hub/0] ocm-controller/start completed in 33.08 seconds
2023-03-30 20:05:16,959 INFO    [hub/0] Running cert-manager/start
2023-03-30 20:05:46,228 INFO    [hub/0] cert-manager/start completed in 29.27 seconds
2023-03-30 20:05:46,228 INFO    [hub/0] Running olm/start
2023-03-30 20:05:56,076 INFO    [dr1/0] rook-cluster/start completed in 196.42 seconds
2023-03-30 20:05:56,077 INFO    [dr1/0] Running rook-pool/start
2023-03-30 20:06:27,468 INFO    [dr1/0] rook-pool/start completed in 31.39 seconds
2023-03-30 20:06:27,468 INFO    [dr1/0] Running rook-toolbox/start
2023-03-30 20:06:55,054 INFO    [hub/0] olm/start completed in 68.83 seconds
2023-03-30 20:08:09,244 INFO    [dr1/2] olm/start completed in 458.64 seconds
2023-03-30 20:08:09,244 INFO    [dr1/2] Running minio/start
2023-03-30 20:08:16,433 ERROR   [dr2/0] Cluster failed
Traceback (most recent call last):
  File "/home/nsoffer/src/ramen/test/drenv/__main__.py", line 111, in execute
    f.result()
  File "/usr/lib64/python3.11/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/usr/lib64/python3.11/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nsoffer/src/ramen/test/drenv/__main__.py", line 264, in run_worker
    run_script(script, worker["name"], hooks=hooks, allow_failure=allow_failure)
  File "/home/nsoffer/src/ramen/test/drenv/__main__.py", line 279, in run_script
    run_hook(hook, script["args"], name, allow_failure=allow_failure)
  File "/home/nsoffer/src/ramen/test/drenv/__main__.py", line 286, in run_hook
    run(hook, *args, name=name)
  File "/home/nsoffer/src/ramen/test/drenv/__main__.py", line 301, in run
    for line in commands.watch(*cmd):
  File "/home/nsoffer/src/ramen/test/drenv/commands.py", line 127, in watch
    raise Error(args, p.returncode, error)
drenv.commands.Error: Command failed:
   command: ('rook-cluster/start', 'dr2')
   exitcode: 1
   error:
      Traceback (most recent call last):
        File "/home/nsoffer/src/ramen/test/rook-cluster/start", line 50, in <module>
          wait(cluster)
        File "/home/nsoffer/src/ramen/test/rook-cluster/start", line 34, in wait
          kubectl.wait(
        File "/home/nsoffer/src/ramen/test/drenv/kubectl.py", line 79, in wait
          _watch("wait", *args, context=context)
        File "/home/nsoffer/src/ramen/test/drenv/kubectl.py", line 95, in _watch
          for line in commands.watch(*cmd, input=input):
        File "/home/nsoffer/src/ramen/test/drenv/commands.py", line 127, in watch
          raise Error(args, p.returncode, error)
      drenv.commands.Error: Command failed:
         command: ('kubectl', 'wait', '--context', 'dr2', 'cephcluster/my-cluster', '--for=jsonpath={.status.phase}=Ready', '--namespace=rook-ceph', '--timeout=300s')
         exitcode: 1
         error:
            error: timed out waiting for the condition on cephclusters/my-cluster

2023-03-30 20:08:17,306 ERROR   [dr1/1] Cluster failed
Traceback (most recent call last):
  File "/home/nsoffer/src/ramen/test/drenv/__main__.py", line 111, in execute
    f.result()
  File "/usr/lib64/python3.11/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/usr/lib64/python3.11/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nsoffer/src/ramen/test/drenv/__main__.py", line 264, in run_worker
    run_script(script, worker["name"], hooks=hooks, allow_failure=allow_failure)
  File "/home/nsoffer/src/ramen/test/drenv/__main__.py", line 279, in run_script
    run_hook(hook, script["args"], name, allow_failure=allow_failure)
  File "/home/nsoffer/src/ramen/test/drenv/__main__.py", line 286, in run_hook
    run(hook, *args, name=name)
  File "/home/nsoffer/src/ramen/test/drenv/__main__.py", line 301, in run
    for line in commands.watch(*cmd):
  File "/home/nsoffer/src/ramen/test/drenv/commands.py", line 127, in watch
    raise Error(args, p.returncode, error)
drenv.commands.Error: Command failed:
   command: ('ocm-cluster/start', 'dr1', 'hub')
   exitcode: 1
   error:
      Traceback (most recent call last):
        File "/home/nsoffer/src/ramen/test/ocm-cluster/start", line 132, in <module>
          deploy(cluster_name, hub_name)
        File "/home/nsoffer/src/ramen/test/ocm-cluster/start", line 48, in deploy
          join_cluster(cluster, hub)
        File "/home/nsoffer/src/ramen/test/ocm-cluster/start", line 92, in join_cluster
          clusteradm.join(
        File "/home/nsoffer/src/ramen/test/drenv/clusteradm.py", line 61, in join
          _watch(*cmd)
        File "/home/nsoffer/src/ramen/test/drenv/clusteradm.py", line 95, in _watch
          for line in commands.watch(*cmd):
        File "/home/nsoffer/src/ramen/test/drenv/commands.py", line 127, in watch
          raise Error(args, p.returncode, error)
      drenv.commands.Error: Command failed:
         command: ('clusteradm', 'join', '--hub-token', 'eyJhbGciOiJSUzI1NiIsImtpZCI6ImRta3IwT3dIY0cwa0c0emxKSVZ6TFBzMEVmVlc3dTdGb1RXNHppdmZjSzQifQ.eyJhdWQiOlsiaHR0cHM6Ly9rdWJlcm5ldGVzLmRlZmF1bHQuc3ZjLmNsdXN0ZXIubG9jYWwiXSwiZXhwIjoxNjgwMTk5Mzk1LCJpYXQiOjE2ODAxOTU3OTUsImlzcyI6Imh0dHBzOi8va3ViZXJuZXRlcy5kZWZhdWx0LnN2Yy5jbHVzdGVyLmxvY2FsIiwia3ViZXJuZXRlcy5pbyI6eyJuYW1lc3BhY2UiOiJvcGVuLWNsdXN0ZXItbWFuYWdlbWVudCIsInNlcnZpY2VhY2NvdW50Ijp7Im5hbWUiOiJjbHVzdGVyLWJvb3RzdHJhcCIsInVpZCI6IjFlMjBlMmU0LTY0NTktNDdiNy1hZDU5LTMwZDM3Y2JhNGZiYiJ9fSwibmJmIjoxNjgwMTk1Nzk1LCJzdWIiOiJzeXN0ZW06c2VydmljZWFjY291bnQ6b3Blbi1jbHVzdGVyLW1hbmFnZW1lbnQ6Y2x1c3Rlci1ib290c3RyYXAifQ.bi1vQBsrIPbiQ5Lgs0rzUIYoH72ikIcAfyJdWWs_Lv1VRZGEkcM7Q_d3uxOSWJVHbnIA5BM1jTHkP0S0_BzS2OfJbAJh805S-ROVRRZHynJcNcg-8um857gmg8866OPs3bYu0n6rgL2rY1FNzGda1VbIQfP7ZFocgFZyonHcrkuXBZ2SD30ST164Yzb1jhGVufvwoySXmzWXghXF9M9I1XcRJfKdjiwzvh1nkQLpisovz8FSUiW2zs40TG6zQmtxdUbs0ZarLCOB7BPko5fjG1JV7zTMrMiZ8Rx940v-pkoazlxMXWOAJ2djA5BmPg_iCCDRfapUFAj_bNFFJkYgQg', '--hub-apiserver', 'https://192.168.122.114:8443', '--cluster-name', 'dr1', '--wait', '--context', 'dr1')
         exitcode: 1
         error:
            E0330 20:03:15.319041  118898 memcache.go:196] couldn't get resource list for packages.operators.coreos.com/v1: the server is currently unable to handle the request
            E0330 20:03:15.321224  118898 memcache.go:101] couldn't get resource list for packages.operators.coreos.com/v1: the server is currently unable to handle the request
            Error: unexpected watch event received

2023-03-30 20:08:46,158 ERROR   [dr2/2] Cluster failed
Traceback (most recent call last):
  File "/home/nsoffer/src/ramen/test/drenv/__main__.py", line 111, in execute
    f.result()
  File "/usr/lib64/python3.11/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/usr/lib64/python3.11/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nsoffer/src/ramen/test/drenv/__main__.py", line 264, in run_worker
    run_script(script, worker["name"], hooks=hooks, allow_failure=allow_failure)
  File "/home/nsoffer/src/ramen/test/drenv/__main__.py", line 279, in run_script
    run_hook(hook, script["args"], name, allow_failure=allow_failure)
  File "/home/nsoffer/src/ramen/test/drenv/__main__.py", line 286, in run_hook
    run(hook, *args, name=name)
  File "/home/nsoffer/src/ramen/test/drenv/__main__.py", line 301, in run
    for line in commands.watch(*cmd):
  File "/home/nsoffer/src/ramen/test/drenv/commands.py", line 127, in watch
    raise Error(args, p.returncode, error)
drenv.commands.Error: Command failed:
   command: ('olm/start', 'dr2')
   exitcode: 1
   error:
      Traceback (most recent call last):
        File "/home/nsoffer/src/ramen/test/olm/start", line 82, in <module>
          wait(cluster)
        File "/home/nsoffer/src/ramen/test/olm/start", line 58, in wait
          kubectl.wait(
        File "/home/nsoffer/src/ramen/test/drenv/kubectl.py", line 79, in wait
          _watch("wait", *args, context=context)
        File "/home/nsoffer/src/ramen/test/drenv/kubectl.py", line 95, in _watch
          for line in commands.watch(*cmd, input=input):
        File "/home/nsoffer/src/ramen/test/drenv/commands.py", line 127, in watch
          raise Error(args, p.returncode, error)
      drenv.commands.Error: Command failed:
         command: ('kubectl', 'wait', '--context', 'dr2', 'csv/packageserver', '--namespace=olm', '--for=jsonpath={.status.phase}=Succeeded', '--timeout=300s')
         exitcode: 1
         error:
            error: timed out waiting for the condition on clusterserviceversions/packageserver

2023-03-30 20:08:57,388 INFO    [dr1/0] rook-toolbox/start completed in 149.92 seconds
2023-03-30 20:12:19,942 INFO    [dr1/2] minio/start completed in 250.70 seconds
2023-03-30 20:12:19,943 INFO    [dr1/2] Running velero/start
2023-03-30 20:12:21,710 INFO    [dr1/2] velero/start completed in 1.77 seconds
2023-03-30 20:12:21,710 INFO    [dr1/2] Running volsync/start
2023-03-30 20:12:24,048 INFO    [dr1/2] volsync/start completed in 2.34 seconds
2023-03-30 20:14:25,810 INFO    [dr2/1] ocm-cluster/start completed in 871.19 seconds
2023-03-30 20:14:25,810 INFO    [dr2/1] Running ocm-cluster/test
2023-03-30 20:14:33,842 INFO    [dr2/1] ocm-cluster/test completed in 8.03 seconds
```

**Comment by nirs** [comment]
We can avoid this by exiting the process on the first error, but this will not clean
up properly in scripts, leaving temporary files and possibly other leftovers.

The best way is to add cancellation support - once one worker fail, set a cancel 
global flag, and wait until all workers are canceled.

**Comment by nirs** [comment]
Fixed in #1241
